### ISSUE TITLE
Proposal: Timer static Create methods that make rooting behavior explicit

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#32866]
**Background**

The rooting behavior employed by `Timer` isn't obvious, and can lead to unexpected leaks like those in https://github.com/dotnet/corefx/pull/32793 and https://github.com/dotnet/corefx/pull/32794.

When you create a `Timer` with `new Timer(delegate, state, dueTime, period)`, the resulting `Timer` object that gets created isn't rooted by the implementation, so if all references to it are dropped, a finalizer will end up canceling the timer. However, the implementation does root the `delegate` and `state` objects as long as the timer is scheduled; when the `Timer` gets dropped and the finalizer unschedules the underlying object, the `delegate` and `state` will be unrooted... however, if either the `delegate` or the `state` reference the `Timer` object, they will keep that from happening.  This kind of cycle happens in a situations like:
```C#
var s = new State();
s.Timer = new Timer(delegate, s, dueTime, period);
```
in which case this `Timer` will end up being rooted by the rooted `State` object until the timer is explicitly unscheduled via a call to `Change(Timeout.Infinite, Timeout.Infinite)`.

Further, when the `new Timer(delegate)` constructor is used, the implementation effectively does the equivalent of `Timer t = new Timer(delegate); t.State = t; t.Change(dueTime, period);` such that the Timer implicitly keeps itself rooted as long as it's scheduled.

There is a non-trivial amount of code that depends on these behaviors, either to keep the `Timer` rooted because that's what's needed (e.g. if you want to create a fire-and-forget timer where it'll go off at a particular due time but without you needing to explicitly keep the Timer rooted), or to ensure it's not rooted (e.g. to have a periodic timer that'll fire only as long as other state is alive and to go away when it's not).  But keeping track of which is which can be difficult.

@davidfowl and I have been discussing ways to address this.  Changing the existing behavior would be a huge breaking change, and since there are actually multiple modes that are desirable at different times, changing to one or the other wouldn't actually fix things.

**Proposal**

The issue https://github.com/dotnet/corefx/issues/26523 suggests adding the API:
```C#
public static Timer UnsafeCreate(TimerCallback callback, object state, TimeSpan dueTime, TimeSpan period);
```
in order to be able to avoid flowing ExecutionContext. I suggest instead we do this:
```C#
public enum TimerLifetimeMode
{
    RootedState, // Timer won't be rooted by the implementation, but delegate/object state will as long as it's scheduled, same as the existing Timer(delegate, state, dueTime, period) ctor
    Rooted, // Timer will be rooted as long as it's scheduled, same as the existing Timer(delegate) ctor
    Unrooted, // Neither Timer nor the TimerCallback nor the object state will be rooted
}
...
public static Timer Create(TimerCallback callback, object state, TimeSpan dueTime, TimeSpan period, TimerLifetimeMode mode = TimerLifetimeMode.RootedState);
public static Timer UnsafeCreate(TimerCallback callback, object state, TimeSpan dueTime, TimeSpan period, TimerLifetimeMode mode = TimerLifetimeMode.RootedState);
```
(There may of course be better names for the enum and enum values.)

That way:
- All of the interesting modes are available and called out in the API / IntelliSense.
- You have all of the options available to you whether you want to flow ExecutionContext or not.
- The default when using {Unsafe}Create still matches the equivalent ctor with the same number of required arguments.

Note that while `Unrooted` would avoid having the implementation maintain rooted references to the `Timer`, delegate, and state, it would still maintain a rooted reference to any captured `ExecutionContext`, as it's extremely unlikely the caller would be maintaining that and it's a critical part of the ability to invoke the callback, so there's still the rare but possible situation where some `AsyncLocal` in that `ExecutionContext` references the `Timer`, in which case it would still be kept alive.  I don't see a good way around that, other than using `UnsafeCreate` to not capture the `ExecutionContext`.

Also, implementation-wise, I expect this would likely end up increasing the size of the objects involved by a few fields, and maybe using `WeakReference`s in the implementation, and maybe adding some casting when firing timers.  For example, the `Rooted` mode could be achieved easily just by adding a `Timer` field to the internal `TimerQueueTimer` object.  But to implement `Unrooted`, `TimerQueueTimer` will no longer be able to store a strong reference to the delegate or object state as it currently does, so for that mode, it would probably need a field that could store either a `WeakReference<DelegateAndState>` (or one for each object) or a `WeakReference<Timer>`, then adding the delegate and object as fields to the `Timer`, or potentially use a `WeakReference<TimerQueueTimer>` for the whole object, keeping it alive only by a `_timerQueueTimer` field on the `Timer`.  Or some such thing.  I don't think it would be prohibitive, but we'd want to do some perf testing to confirm before going ahead with it if/when the APIs are approved.

cc: @kouvel, @davidfowl, @vancem

**Comment by GSPP** [comment]
Maybe `RootedState` should be markted `[Obsolete]` immediately. It seems to be a compatibility mode that effectively acts as either `Rooted` or `Unrooted` depending on how calling code is written.

**Comment by vancem** [comment]
I note a few things

1. The unrooted behavior is strictly harder than the other two.  We have to go out of our way to have weak handles in various places, so the code I strictly more complex/expensive.
2. It looks like you could build the Unrooted implementation 'on top' of a rooted one.   (This is how you 'fixed' the leaks you found.   Actually structuring the implementation this way, would be very good since it is pay for play (if I don't need the unrooted behavior I don't have to pay for it.
3. My expectation is that most 'normal' uses of timer would have an object that holds it, and that the callback is VERY likely to reference that object, forming the reference loop until the timer fires.  The times that there is a interesting difference between StateRooted and Rooted is probably pretty rare (probably when you create a timer and never assign it to any field), and feel like bugs (the target obejct lose the ability to control cancel/control whether  timer fires or not (it depends on when the Finalizer happens with respect to the timer).     This does argue for @GSPP s suggestion of making it obsolete (suggesting Rooted).  

Finally I note that the alternative to this proposal is to simply more carefully document the API, and provide good guidance.   The leaks that prompted this proposal don't get 'fixed' by this change, they just get a bit easier to implement (since Timer supports the Unrooted (might want change the terminology from 'Root/Unrooted' to 'Strong/Weak') directly.  

In general, however I don't see a problem with the proposal.   I would like to insure that the implementation is pay-for-play (the common 'Rooted' case does not pay for the 'Unooted' cased)

**Comment by stephentoub** [comment]
> The unrooted behavior is strictly harder than the other two.

Yes, definitely.  It's both more complicated and involves more cost, since the way we represent weak references involves more cost.

> It looks like you could build the Unrooted implementation 'on top' of a rooted one.

I'm not convinced that's true, at least not without additional changes.  It worked for the cases I found because the state in question was already being kept alive elsewhere, but that's not necessarily going to be the case for all Unrooted usage.  If we just changed it to use a weak reference to wrap the delegate and the state, we might lose those references even if the Timer itself is kept alive elsewhere.  As such, I think we'd need to augment Timer itself to maintain the delegate and state references, which would either mean we increase the size of Timer, or we unseal it and use a derived type to store the references.

> My expectation is that most 'normal' uses of timer would have an object that holds it, and that the callback is VERY likely to reference that object, forming the reference loop until the timer fires

I agree.

**Comment by vancem** [comment]
> It looks like you could build the Unrooted implementation 'on top' of a rooted one.

> I'm not convinced that's true.

The 'on top of' suggestion was just a thought.  I just want it to be simple and pay for play (no weak handles when we don't need them).   I note that you coudl use conditional weak handles, but that is probably overkill.   You basically need to have the state and callback referenced from Timer (rather than the queue entry), and have a weak handle back to it.  I could live with the extra pointer (or two) in Timer (that is unused in the non-weak case).   The rest could live 'on top' (the TimerQueue holds a delgate that looks at a weak handle back to the timer to fire it if it is still alive).    Again the actual implementor gets to fiddle with the details, I just want it to be pay-for-play (and not too complicated).

**Comment by stephentoub** [comment]
>  Again the actual implementor gets to fiddle with the details, I just want it to be pay-for-play (and not too complicated).

Yup, agreed.  And, yeah, you could use a ConditionalWeakTable to avoid expanding Timer by a few fields.

**Comment by terrajobst** [comment]
Please first reconcile with  #26523 and decide on an approach.

**Comment by stephentoub** [comment]
#26523 is a subset of this proposal (the only difference between that API and the corresponding one here is that this one has an additional optional argument).

**Comment by terrajobst** [comment]
Given the problematic design of the existing API it seems we don't want the first enum value, which means we'd get away with making the `bool`. We should default them to sane behavior, which is to root the timer.

```C#
public static Timer Create(TimerCallback callback, object state, TimeSpan dueTime, TimeSpan period, bool unrooted = false);
public static Timer UnsafeCreate(TimerCallback callback, object state, TimeSpan dueTime, TimeSpan period, bool unrooted = false);
```

**Comment by GSPP** [comment]
The first enum value is there to specify the existing behavior. The existing behavior must be preserved for compatibility.

The first enum member could be removed but then the new constructors no longer offer a superset of the functionality of the old ones. That causes confusion. Keeping the enum value does not do much harm since the behavior must be supported anyway.

**Comment by benaadams** [comment]
> The first enum value is there to specify the existing behavior. The existing behavior must be preserved for compatibility.

The existing behaviour is undefined; sometimes would root sometimes would not. It can be achieved by using the existing .`ctor`s as they aren't going away.

**Comment by GSPP** [comment]
It does not matter whether it's documented or not. For compatibility it is set in stone. In fact it *should* be documented now.

**Comment by benaadams** [comment]
>  For compatibility it is set in stone. 

Not sure I understand. The existing .ctor behaviour is not changing; these are completely new methods that no one is currently calling. How does compatibility feature?

**Comment by GSPP** [comment]
@benaadams I understand your point, but did you see this?

> The first enum member could be removed but then the new constructors no longer offer a superset of the functionality of the old ones. That causes confusion. Keeping the enum value does not do much harm since the behavior must be supported anyway.

It seems confusing to have to tell API users that they should not use the old constructors anymore when at the same time the new ones are not a superset of the old ones.

If we make the new ones a functional superset then all constructors are homogenous. They are just different overloads over the same functionality. This always seemed like a best practice for overloading to me.

Doing that also helps with API discoverability. When a user is new to the timer class and wants to understand what options he has, he will discover the enum and discover all the options. He will not fall into the trap of using the old constructor and thereby getting crazy behavior that is not apparent anywhere in the API surface.

**Comment by stephentoub** [comment]
From my perspective, either:
a) We have the enum with all three modes, and we either don't have a default mode for Create (the parameter is non-optional) or we default to the same behavior as the ctor with the equivalent arguments.  We'd choose this approach if we want to make it explicitly clear in the new APIs that the Timer is capable of three different modes of operation, in hopes that even though we don't expect developers to want the "rooted delegate/state" behavior, the API will hopefully teach them what the differences are, make them aware of the differences, etc.
b) We don't have the enum, and just have a bool.  We'd choose this approach if we want to push developers to use Create instead of the ctor, and don't want to carry the old/odd behavior forward to these new methods, leaving the behavior as an oddity of the ctors (which, I agree, should be documented if it's not already), at which point there's little benefit to adding an additional type.

(There's a (c) I don't like, which is to add an enum that just has Rooted/Unrooted as two options.  At that point, we should just do (b).)

I don't have a strong preference between (a) and (b).  I proposed (a) for the reasons described, but in the API review folks were skeptical of bringing forward the old behavior.  @KrzysztofCwalina, you were leading that charge... strong preference?

**Comment by benaadams** [comment]
The various options were discussed on the review https://youtu.be/1H3f1DbB_N8?t=3316 (from 55:16 to 1:19:30) 

The bool in the `Create` method offers a specific behaviour; from what was explained in the review the `enum` would be more like:
```csharp
public enum TimerLifetimeMode
{
    // Current behaviour: Indeterminate
    // Depends if is timer stored in field; what the delegate captures; 
    // what other methods are called on timer and the timer period
    Indeterminate, 
    Rooted,  // Always fires while scheduled, keeps objects alive to run timer (prevents GC on them)
    Unrooted // Dechedules if the objects holding the timer deferenced and GC'd (allows GC on them)
}
```

**Comment by stephentoub** [comment]
> Indeterminate

The behavior is determinate: the implementation roots the delegate and the state object.  The challenge for the enum was coming up with a good name for that.

> The bool in the Create method offers a specific behaviour

I'm not sure what this means... the bool just maps to the Rooted/Unrooted behavior from the enum.

**Comment by benaadams** [comment]
> > Indeterminate

> The behavior is determinate: the implementation roots the delegate and the state object. 

And what that means cannot be determined by the method signature alone; it depends on the delegate and `state` and whether they have a reference to the Timer.

> > The bool in the Create method offers a specific behaviour

> I'm not sure what this means... 

I mean the method signiture is enough to determine what will happen without considering what is actually in `state` or what the delegate captures

**Comment by stephentoub** [comment]
> And what that means cannot be determined by the method signature alone

Sure it can: the delegate and the state will be rooted.  As with everything else in .NET, anything they reference would be rooted, too.  Same goes for the resulting Timer keeping alive AsyncLocals, which aren't "in the signature".

**Comment by benaadams** [comment]
Well, yes :) Just mean its more explicit

Though not keeping AsyncLocals alive is in the signature via `Unsafe`

**Comment by stephentoub** [comment]
> Though not keeping AsyncLocals alive is in the signature via Unsafe

Different method ;)

**Comment by vancem** [comment]
Responding to @stephentoub  options (a) and (b) above.   

First I dont view the current behavior as that bad.   From a practical standpoint, either 

1. I am signed up to explicitly dispose of the timer when I no longer need it OR
2.  I am relying on the GC to figure out when the timer dies and turn it off then.  

If you are in case (1) which should be the common case, the differnece between the current behavior and 'rooted' is probably immaterial.   Thus I don't think we should be agressively moving people away from it (it works fine, we have bigger fish to fry).     It only real disadvantage (over rooted), is that it is harder to describe (but in the common case you don't care).  

Thus I don't think we should be agressively deprecating current behavior (it is fine), which in turn suggests that Stephen's option (a) (his original proposal, is also fine.     A documentation fix is the main thing (which includes prefernece to using true Rooted, and a recomendation to use the new unrooted option if you are doing (2).   

For what it is worth....

**Comment by stephentoub** [comment]
Moving back to api-ready-to-review to discuss again...

**Comment by terrajobst** [comment]
We discussed this and we feel that the only argument in favor of the proposal is accomodate people that want to use `UnsafeCreate` but want the existing constructor behavior, we seems far fetched, considering how nuanced the behavior is. We believe virtually everyone will want either rooted or non-rooted. My personal opinion is that rooted is probably what most people would expect, so I'd make this the default. We probably want to avoid having to add a new type, considering that it's only two states. And if we ever have to, we can still add the enum. Which brings us back to the proposal we approved last week.

**Comment by vancem** [comment]
Just for clarity, what is the proposal you approved last week? (I am guessing it is (b) in Stephen's comment above (a bool on the Create and UnsafeCreate APIs mentioned above
```
public static Timer Create(TimerCallback callback, object state, TimeSpan dueTime, TimeSpan period, bool rooted = true);
public static Timer UnsafeCreate(TimerCallback callback, object state, TimeSpan dueTime, TimeSpan period, bool rooted = true);
```
